src/ClientData/ThreadTrackPaneData.cpp

```diff
@@ -0,0 +1,107 @@
+  return scope_tree_.Height() - 1;
+}
+
+const orbit_client_protos::TimerInfo& ThreadTrackPaneData::AddTimer(
+    orbit_client_protos::TimerInfo timer_info) {
+  const auto& timer_info_chain_ref = track_pane_data_.AddTimer(/*depth=*/0, timer_info);
+  UpdateMinTime(timer_info.start());
+  UpdateMaxTime(timer_info.end());
+
+  if (scope_tree_update_type_ == ScopeTreeUpdateType::kAlways) {
+    absl::MutexLock lock(&scope_tree_mutex_);
+    scope_tree_.Insert(&timer_info_chain_ref);
+  }
+  return timer_info_chain_ref;
+}
+
+void ThreadTrackPaneData::BuildScopeTreeFromTrackData() {
+  if (scope_tree_update_type_ != ScopeTreeUpdateType::kOnCaptureComplete) {
+    return;
+  }
+  std::vector<const TimerChain*> timer_chains = track_pane_data_.GetChains();
+  for (const TimerChain* timer_chain : timer_chains) {
+    CHECK(timer_chain != nullptr);
+    absl::MutexLock lock(&scope_tree_mutex_);
+    for (const auto& block : *timer_chain) {
+      for (size_t k = 0; k < block.size(); ++k) {
+        scope_tree_.Insert(&block[k]);
+      }
+    }
+  }
+}
+
+std::vector<const orbit_client_protos::TimerInfo*> ThreadTrackPaneData::GetTimers(
+    uint64_t min_tick, uint64_t max_tick) const {
+  std::vector<const orbit_client_protos::TimerInfo*> all_timers;
+  absl::MutexLock lock(&scope_tree_mutex_);
+
+  for (const auto& [depth, ordered_nodes] : scope_tree_.GetOrderedNodesByDepth()) {
+    // Special case. ScopeTree has a root node in depth 0 which shouldn't be considered.
+    if (!depth) continue;
+
+    auto first_node_to_draw = ordered_nodes.upper_bound(min_tick);
+    if (first_node_to_draw != ordered_nodes.begin()) --first_node_to_draw;
+
+    // If this node is strictly before the range, we shouldn't include it.
+    if (first_node_to_draw->second->GetScope()->end() < min_tick) ++first_node_to_draw;
+
+    for (auto it = first_node_to_draw; it != ordered_nodes.end() && it->first < max_tick; ++it) {
+      all_timers.push_back(it->second->GetScope());
+    }
+  }
+  return all_timers;
+}
+
+const orbit_client_protos::TimerInfo& ThreadTrackPaneData::GetLeft(
+    const orbit_client_protos::TimerInfo& timer) const {
+  absl::MutexLock lock(&scope_tree_mutex_);
+  return *scope_tree_.FindPreviousScopeAtDepth(timer);
+}
+
+const orbit_client_protos::TimerInfo& ThreadTrackPaneData::GetRight(
+    const orbit_client_protos::TimerInfo& timer) const {
+  absl::MutexLock lock(&scope_tree_mutex_);
+  return *scope_tree_.FindNextScopeAtDepth(timer);
+}
+
+const orbit_client_protos::TimerInfo& ThreadTrackPaneData::GetUp(
+    const orbit_client_protos::TimerInfo& timer) const {
+  absl::MutexLock lock(&scope_tree_mutex_);
+  return *scope_tree_.FindParent(timer);
+}
+
+const orbit_client_protos::TimerInfo& ThreadTrackPaneData::GetDown(
+    const orbit_client_protos::TimerInfo& timer) const {
+  absl::MutexLock lock(&scope_tree_mutex_);
+  return *scope_tree_.FindFirstChild(timer);
+}
+
+void ThreadTrackPaneData::UpdateMinTime(uint64_t min_time) {
```

Comment:
is this different from min/max time in TrackPaneData?

Reply:
Not, exactly the same. In a following PR I was thinking to use ScopeTree to get the min/max time without needing these 2 members.

Reply:
I still think you can use min/max from track_pane_data_. I do not see the difference.

Reply:
Yes, sorry. I will change it.